File: JavaScript/425-Word Squares.js
/**
 * @param {string[]} words
 * @return {string[][]}
 */
var wordSquares = function(words) {
    const N = words[0].length;
    let prefixMap = new Map();
    for (let word of words){
        for (let i=1; i<N; i++){
            const w = word.substring(0, i);
            if (prefixMap.has(w)){
                prefixMap.get(w).push(word);
            }else{
                prefixMap.set(w, new Array(1).fill(word));
            }
        }
    }
    
    let res = new Array();
    
    let backtrack = (idx, curr) =>{
        if (curr.length == words[0].length){
            res.push(curr);
            return;
        }
        let prefix = "";
        for (let prev of curr){
            prefix += prev[idx];
        }
        if (prefixMap.has(prefix)){
            for (let nxt of prefixMap.get(prefix)){
                backtrack(idx+1, curr.concat([nxt]));
            }
        }
    }
    
    for (let word of words){
        backtrack(1, new Array(1).fill(word));
    }
    return res;
};

File: C++/425-Word Squares.cpp
// O(N × 26^L), where L is the word length (1 ≤ L ≤ 4)
// 26^L represents all possible lowercase alphabet combinations of length L
class Solution {
public:
    int N;
    vector<vector<string>> res;
    void dfs(int cnt, vector<string>& curr, vector<string>& words, unordered_map<string, vector<int>>& prefixMap) {
        if (cnt == N) {
            res.push_back(curr);
            return;
        }

        string prefix = "";
        for (int i=0; i<curr.size(); i++) {
            prefix.push_back(curr[i][cnt]);
        }

        for (int wordIdx:prefixMap[prefix]) {
            curr.push_back(words[wordIdx]);
            dfs(cnt+1, curr, words, prefixMap);
            curr.pop_back();
        }

    }
    vector<vector<string>> wordSquares(vector<string>& words) {
        unordered_map<string, vector<int>> prefixMap;
        this->N = words[0].size();
        for (int i=0; i<words.size(); i++) {
            for (int j=0; j<N; j++) {
                string w = words[i].substr(0, j);
                prefixMap[w].push_back(i);
            }
        }

        for (int i=0; i<words.size(); i++) {
            vector<string> arr = {words[i]};
            dfs(1, arr, words, prefixMap);
        }

        return res;
    }
};